types/src/v0/header.rs

```diff
@@ -1,20 +1,166 @@
+            }
+            .serialize(serializer),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Header {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct HeaderVisitor;
+
+        #[derive(Deserialize)]
+        #[serde(field_identifier, rename_all = "lowercase")]
+        enum StructFields {
+            Version,
+            Fields,
+        }
+
+        impl<'de> Visitor<'de> for HeaderVisitor {
+            type Value = Header;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("Header")
+            }
+
+            fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
+            where
+                V: SeqAccess<'de>,
+            {
+                let chain_config_or_version: ResolvableChainConfigOrVersion =
+                    seq.next_element()?.unwrap();
+                match chain_config_or_version.chain_config {
+                    // For v0.1, the first field in the sequence of fields is the first field of the struct, so we call a function to get the rest of
+                    // the fields from the sequence and pack them into the struct.
+                    EitherOrVersion::Left(cfg) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config(cfg.into(), seq)?,
+                    )),
+                    EitherOrVersion::Right(commit) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config(commit.into(), seq)?,
+                    )),
+                    // For all versions, the first "field" is not actually part of the `Header` struct (you can think of it as the first field of the virtual
+                    // `VersionedHeader` struct, with the `Header` being the second field. We just delegate directly to the derived deserialization
+                    // impl for the appropriate version.
+                    EitherOrVersion::Version(Version { major: 0, minor: 2 }) => Ok(Header::V2(
+                        seq.next_element()?
+                            .ok_or_else(|| de::Error::missing_field("fields"))?,
+                    )),
+                    EitherOrVersion::Version(Version { major: 0, minor: 3 }) => Ok(Header::V3(
+                        seq.next_element()?
+                            .ok_or_else(|| de::Error::missing_field("fields"))?,
+                    )),
+                    _ => Err(serde::de::Error::custom("invalid version")),
+                }
+            }
+
+            fn visit_map<V>(self, mut map: V) -> Result<Header, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                // retrieving the first entry from the map, which could either be a
+                // `chain_config` (for v0.1 headers) or a `version` (for `VersionedHeader` structs).
+                let Some((_, chain_config_or_version)) =
```

Comment:
I am assuming that first field is always chain_config or version but that might not be the case.

Reply:
Hmm yeah. The best way might be to grab all the entries and collect into a `HashMap`. Then we can look up specific keys like `version` or `chain_config`